daemon/containerd/image_delete.go

```diff
@@ -53,123 +54,180 @@ import (
 //
 // TODO(thaJeztah): image delete should send prometheus counters; see https://github.com/moby/moby/issues/45268
 func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force, prune bool) ([]imagetypes.DeleteResponse, error) {
-	parsedRef, err := reference.ParseNormalizedNamed(imageRef)
-	if err != nil {
-		return nil, err
+	var c conflictType
+	if !force {
+		c |= conflictSoft
 	}
 
-	img, err := i.resolveImage(ctx, imageRef)
+	img, all, err := i.resolveAllReferences(ctx, imageRef)
 	if err != nil {
 		return nil, err
 	}
 
-	imgID := image.ID(img.Target.Digest)
-
-	explicitDanglingRef := strings.HasPrefix(imageRef, imageNameDanglingPrefix) && isDanglingImage(img)
-	if isImageIDPrefix(imgID.String(), imageRef) || explicitDanglingRef {
-		return i.deleteAll(ctx, img, force, prune)
-	}
-
-	singleRef, err := i.isSingleReference(ctx, img)
-	if err != nil {
-		return nil, err
-	}
-	if !singleRef {
-		err := i.client.ImageService().Delete(ctx, img.Name)
+	var imgID image.ID
+	if img == nil {
+		if len(all) == 0 {
+			return []imagetypes.DeleteResponse{}, nil
+		}
+		imgID = image.ID(all[0].Target.Digest)
+		var named reference.Named
+		if imageRef != string(imgID) {
```

Comment:
Shouldn't this also check for truncated ID (and the dangling image)?
Like here:
https://github.com/moby/moby/blob/5c48d187eeb356f5617a3ffda5adfcfef5d5b986/daemon/containerd/image_delete.go#L107-L108

If so, we should probably extract that to a function an call it here.

Reply:
Could check not a prefix, no need to check dangling since it would already hit this branch as not a hash or hash prefix.

Reply:
Actually just `!isImageIDPrefix(imgID.String(), imageRef)` is enough here since that covers the equal case already